README.md
<!--
 * @Author: Nick Liu
 * @Date: 2022-03-22 10:53:09
 * @LastEditTime: 2022-06-30 16:15:17
 * @LastEditors: Nick Liu
 * @Description: The readme file
 * @FilePath: /init-network-per-net-change-mac/README.md
-->
# Auto HotSpot

## About

Automatically Connect Mac to personal hotspot if no more preferred network available, based on network change. i.e. turn on/off wifi, connected/disconnected to a network, etc.

### Intro

As of macOS 12.4, Apple does not allow you to automatically initialize connection to your iPhone's hotspot as it treats the hotspot in a proprietary fashion which differs than treating other Wifi networks. This tool helps you achieve that.
This program is intentionally written and will be maintained in the future in pure shell(bash) script so that no dependencies need to be introduced.

### How does it work

When network change like turning on/off Wifi takes place, this tool is triggered and will check for connectivity, available wifis, if trigger ssid is around and so on. If all conditions are met, it will keep trying to connect to your specified hotspot at a specified frequency for a specified duration, until either connected to this hotspot or above trigger condition fails(e.g. already connected to another network).

## Usage

### Requirements

* Elevated privilege to install the daemon
* Grant *Access Control* permission to `/usr/bin/security` of the hotspot ssid in *Keychain Access*
* Password for the hotspot has been saved to *keychain*

### Dependencies

* None

### Install

1. *Grant*Access Control*permission to `/usr/bin/security` of the hotspot ssid in*Keychain Access*
    1. Open *Keychain Access*
    2. Select *All items* tab
    3. Input the SSID of your hotspot in the search box and search
    4. Select the record with of your SSID name with *System* in *Keychain* column and double click
    5. Choose *Access Control* tab
    6. Click + and input the path to security `/usr/bin` (keyboard shortcut: `command shift g`)
    7. select *security* and *Add*
2. clone the repo
3. Modify the `ProgramArguments` key in plist file to your need
    * The first arg is the program itself, DON'T modify
    * The second arg is the SSID of your hotspot
    * The third arg is the maximum time duration to run this script when network change occurs and not yet connected to hotspot
    * The fourth arg is the trigger SSID that if present in available Wifis or unspecified then connect to hotspot, otherwise do nothing.
4. `sudo ./install.sh`

### Uninstall

`sudo ./uninstall.sh`

## Future work

* [ ] Add more trigger conditions

## Authors

* [@bboysnick5](https://github.com/bboysnick5)

## License

[Apache 2.0](<https://www.apache.org/licenses/LICENSE-2.0>)

## Credits

* <https://gist.github.com/albertbori/1798d88a93175b9da00b>
* <https://github.com/paulbhart/toggleairport>


CHANGELOG.md
## [1.0.1](https://github.com/OnDaShouldersOfGiants/auto-hotspot/compare/v1.0.0...v1.0.1) (2022-07-07)


### Bug Fixes

* **install.sh:** remove usage in help menu ([a2b9277](https://github.com/OnDaShouldersOfGiants/auto-hotspot/commit/a2b92770c7e451c4fc491fe38c1adb1a4f504d57))
* **install.sh:** Show usage only when no args ([3f04aa6](https://github.com/OnDaShouldersOfGiants/auto-hotspot/commit/3f04aa6b8b3fd254e5d794c7aaeb47956680e0cc))


### Reverts

* Revert "chore(.releaserc.json): change config location" ([0b63b9c](https://github.com/OnDaShouldersOfGiants/auto-hotspot/commit/0b63b9c211ef3890b5db6853983fc98e9fa0142b))

# 1.0.0 (2022-06-30)


### Bug Fixes

* **install.sh:** display help when no args given ([223036a](https://github.com/OnDaShouldersOfGiants/auto-hotspot/commit/223036a3855a5bc0fddad7a13d7b8440a918d9b1))


auto-hotspot.sh
#!/bin/bash
###
# @Author: Nick Liu
# @Date: 2022-03-22 10:53:09
# @LastEditTime: 2022-07-07 05:45:53
# @LastEditors: Nick Liu
# @Description: Audo connect Mac to hotspot with supplied command line args as config
# @FilePath: /init-network-per-net-change-mac/auto-hotspot.sh
###

trap -- '' SIGINT SIGTERM

function notify() {
    launchctl asuser "$2" osascript -e "display notification \"Connected to hotspot $1\" with title \"Network Connected\""
}

# fix tests
function is_connected() {
    # if connected to cable ethernet
    for eth_name in $1; do
        if ifconfig "$eth_name" | grep -q 'status: active'; then
            true
            return
        fi
    done

    # if connecteed to wifi
    if /usr/sbin/networksetup -getairportnetwork "$2" | grep -q 'You are not associated with an AirPort network.'; then
        false
        return
    else
        true
        return
    fi
}

# get the password for passed-in item
function get_pw() {
    security find-generic-password -ga "$1" 2>&1 >/dev/null | grep "password: " | sed -e "s/^password: \"//" -e "s/\"$//"
}

function try_connect() {
    start_time_stamp=$SECONDS
    prev_connected=false
    connected_by_script=false
    while ((SECONDS - start_time_stamp < $3)) || prev_connected; do
        # make sure connected for at least 10 seconds
        if is_connected "$4" "$5"; then
            if $prev_connected; then
                $connected_by_script
                return
            fi
            prev_connected=true
            sleep 10
            continue
        else
            prev_connected=false
        fi
        # if to check trigger ssid and trigger ssid is not present
        if "$6" && ! /System/Library/PrivateFrameworks/Apple80211.framework/Versions/Current/Resources/airport -s | awk '{print $1}' | tail -n+2 | grep -q "$1"; then
            false
            return
        fi
        if networksetup -setairportnetwork "$5" "$2" "$(get_pw "$2")" -eq 0; then
            connected_by_script=true
        else
            sleep 4
        fi
    done
}

function main() {
    hotspot_ssid="$1"
    # the trigger ssid that allow this script to keep trying to connect to hotspot if present
    trigger_ssid="$2"
    max_retry_duration="$3"
    check_trigger_ssid=true
    if [ -z "$trigger_ssid" ]; then
        check_trigger_ssid=false
    fi

    # names of the network services
    eth_names=$(networksetup -listnetworkserviceorder | sed -En 's/^\(Hardware Port: .*(Ethernet|LAN).* Device: (en[0-9]+)\)$/\2/p')
    air_name=$(networksetup -listnetworkserviceorder | sed -En 's/^\(Hardware Port: (Wi-Fi|AirPort).* Device: (en[0-9]+)\)$/\2/p')

    if try_connect "$trigger_ssid" "$hotspot_ssid" "$max_retry_duration" "$eth_names" "$air_name" "$check_trigger_ssid"; then
        notify "$hotspot_ssid" "$4"
    fi
}

main "$@"


toggleAirport.sh
#!/bin/bash
#git@github.com:paulbhart/toggleairport.git
#originally from https://gist.github.com/albertbori/1798d88a93175b9da00b
function set_airport {

    new_status=$1

    if [ $new_status = "On" ]; then
        /usr/sbin/networksetup -setairportpower $air_name on
        touch /var/tmp/prev_air_on
    else
        /usr/sbin/networksetup -setairportpower $air_name off
        if [ -f "/var/tmp/prev_air_on" ]; then
            rm /var/tmp/prev_air_on
        fi
    fi

}

function growl {

    # Checks whether Growl is installed
    if [ -f "/usr/local/bin/growlnotify" ]; then
        /usr/local/bin/growlnotify -m "$1" -a "AirPort Utility.app"
    else
        osascript -e "display notification \"$1\" with title \"Wifi Toggle\" sound name \"Hero\""
    fi

}

# Set default values
prev_eth_status="Off"
prev_air_status="Off"
eth_status="Off"

# Grab the names of the adapters. We assume here that any ethernet connection name ends in "Ethernet"
eth_names=`networksetup -listnetworkserviceorder | sed -En 's/^\(Hardware Port: .*(Ethernet|LAN).* Device: (en[0-9]+)\)$/\2/p'`
air_name=`networksetup -listnetworkserviceorder | sed -En 's/^\(Hardware Port: (Wi-Fi|AirPort).* Device: (en[0-9]+)\)$/\2/p'`

# Determine previous ethernet status
# If file prev_eth_on exists, ethernet was active last time we checked
if [ -f "/var/tmp/prev_eth_on" ]; then
    prev_eth_status="On"
fi

# Determine same for AirPort status
# File is prev_air_on
if [ -f "/var/tmp/prev_air_on" ]; then
    prev_air_status="On"
fi

# Check actual current ethernet status
for eth_name in ${eth_names}; do
    if ([ "$eth_name" != "" ] && [ "`ifconfig $eth_name | grep "status: active"`" != "" ]); then
        eth_status="On"
    fi
done

# And actual current AirPort status
air_status=`/usr/sbin/networksetup -getairportpower $air_name | awk '{ print $4 }'`

# If any change has occured. Run external script (if it exists)
if [ "$prev_air_status" != "$air_status" ] || [ "$prev_eth_status" != "$eth_status" ]; then
    if [ -f "./statusChanged.sh" ]; then
        "./statusChanged.sh" "$eth_status" "$air_status" &
    fi
fi

# Determine whether ethernet status changed
if [ "$prev_eth_status" != "$eth_status" ]; then

    if [ "$eth_status" = "On" ]; then
        set_airport "Off"
        growl "Wired network detected. Turning AirPort off."
    else
        set_airport "On"
        growl "No wired network detected. Turning AirPort on."
    fi

# If ethernet did not change
else

    # Check whether AirPort status changed
    # If so it was done manually by user
    if [ "$prev_air_status" != "$air_status" ]; then
    set_airport $air_status

    if [ "$air_status" = "On" ]; then
        growl "AirPort manually turned on."
    else
        growl "AirPort manually turned off."
    fi

    fi

fi

# Update ethernet status
if [ "$eth_status" == "On" ]; then
    touch /var/tmp/prev_eth_on
else
    if [ -f "/var/tmp/prev_eth_on" ]; then
        rm /var/tmp/prev_eth_on
    fi
fi

exit 0


auto-connect-hotspot.sh
#!/bin/bash
###
# @Author: Nick Liu
# @Date: 2022-03-22 10:53:09
# @LastEditTime: 2022-04-03 16:07:59
# @LastEditors: Nick Liu
# @Description: Audo connect Mac to hotspot with supplied command line args as config
# @FilePath: /init-network-per-net-change-mac/auto-connect-hotspot.sh
###

function notify() {
    osascript -e 'display notification "Connected to hotspot "' \
        "$1"'" with title "Connected"'
}

function is_connected() {
    # if connected to cable ethernet
    for eth_name in $1; do
        if ifconfig "$eth_name" | grep -q 'status: active'; then
            true
        fi
    done

    # if connecteed to wifi
    if /usr/sbin/networksetup -getairportnetwork "$2" | grep -q 'You are not associated with an AirPort network.'; then
        false
    else
        true
    fi
}

# get the password for passed-in item
function get_pw() {
    security find-generic-password -ga "$1" 2>&1 >/dev/null | grep "password: " | sed -e "s/^password: \"//" -e "s/\"$//"
}

function try_connect() {
    start_time_stamp=$SECONDS
    while ! is_connected "$4" "$5" && ((SECONDS - start_time_stamp < $3)); do
        if "$6" && ! /System/Library/PrivateFrameworks/Apple80211.framework/Versions/Current/Resources/airport -s | awk '{print $1}' | tail -n+2 | grep -q "$1"; then
            break
        fi
        networksetup -setairportnetwork "$5" "$2" "$(get_pw "$2")"
        sleep 4
    done
}

function main() {
    hotspot_to_connect_ssid="$1"
    max_retry_duration="$2"
    # the trigger ssid that allow this script to keep trying to connect to hotspot if present
    trigger_to_connect_ssid="$3"
    check_trigger_ssid=true
    if [ -z "$trigger_to_connect_ssid" ]; then
        check_trigger_ssid=false
    fi

    # names of the network services
    eth_names=$(networksetup -listnetworkserviceorder | sed -En 's/^\(Hardware Port: .*(Ethernet|LAN).* Device: (en[0-9]+)\)$/\2/p')
    air_name=$(networksetup -listnetworkserviceorder | sed -En 's/^\(Hardware Port: (Wi-Fi|AirPort).* Device: (en[0-9]+)\)$/\2/p')

    if is_connected "$eth_names" "$air_name"; then
        return 0
    fi

    try_connect "$trigger_to_connect_ssid" "$hotspot_to_connect_ssid" "$max_retry_duration" "$eth_name" "$air_name" "$check_trigger_ssid"
    if /usr/sbin/networksetup -getairportnetwork "$air_name" | awk '{ print $4 }' | grep -q 'ip'; then
        notify "$hotspot_to_connect_ssid"
    fi
}

main "$@"


install.sh
#!/bin/bash
###
# @Author: Nick Liu
# @Date: 2022-03-22 10:53:09
# @LastEditTime: 2022-07-06 16:06:15
# @LastEditors: Nick Liu
# @Description: the utility script for installation/uninstallation and help menu
# @FilePath: /init-network-per-net-change-mac/install.sh
###

########### Formats     ############

readonly RED="31"
readonly GREEN="32"
readonly YELLOW="33"
readonly BLUE="34"
readonly BOLD="1"
readonly NORMAL="0"
readonly BLACK="30"

########### Menu flags  ############
# first element for getopts
# second element is short option
# third element is long option
readonly HELP_FLAG=("h" "help" "h")
readonly VERSION_FLAG=("v" "version" "v")
readonly HOTSPOT_SSID_FLAG=("s" "hotspot_ssid" "s:" "hotspot_ssid")
readonly TRIGGER_SSID_FLAG=("t" "trigger_ssid" "t:" "trigger_ssid")
readonly RETRY_DURATION_FLAG=("d" "retry_duration" "d:" "retry_duration")
readonly INSTALL_FLAG=("i" "install" "i")
readonly UNINSTALL_FLAG=("u" "uninstall" "u")

# print line
# first param: false: no-op
#              true: empty line
#              otherwise: treat first and the rest params as printf params
fmt_println() {
    flag_start_idx=1
    num_flags=$#-1
    if [[ "$1" == "false" ]]; then
        # no-op
        return
    elif [[ "$1" == "true" ]]; then
        # empty line
        if ((num_flags == 0)); then
            printf "\n"
            return
        fi
        flag_start_idx=2
        ((num_flags -= 1))
    fi

    str=${*:$#} # last arg
    old_ifs="$IFS"
    IFS=';'
    # everything but last arg(and first param when it's true/false)
    flags="'${*:$flag_start_idx:$num_flags}'"
    IFS=$old_ifs
    printf "\e[${flags}m%s\e[m\n" "$str"
}

indent() {
    pr -to "$1"
}

is_in() {
    local es match="$1"
    shift
    for es; do [[ "$es" == "$match" ]] && return 0; done
    return 1
}

version() {
    fmt_println "ver"
}

# TODO: colorful required and optional flag, colorful default/unspecified
# TODO: refactor println with print block
help() {
    # whether to display usage only or full help
    is_short=$([ "$3" == "USAGE" ] && echo "false" || echo "true")
    is_bold_flags=$([ "$3" == "USAGE" ] && echo "$NORMAL" || echo "$BOLD")

    install_flag_desc="[{-${INSTALL_FLAG[0]}|--${INSTALL_FLAG[1]}}"
    install_flag_desc_indent=$((${#install_flag_desc} + 1))
    fmt_println "$BOLD" "$BLUE" "$3:"
    # install
    fmt_println "$is_bold_flags" "$1 $install_flag_desc"
    fmt_println "$is_short" "Install this script" | indent "$2"
    fmt_println "$is_short"
    # hotspot ssid
    fmt_println "$is_bold_flags" "<{-${HOTSPOT_SSID_FLAG[0]}|--${HOTSPOT_SSID_FLAG[1]}} <${HOTSPOT_SSID_FLAG[3]}>>" | indent $(($2 + install_flag_desc_indent))
    fmt_println "$is_short" "The personal hotspot ssid to try to connect to" | indent $(($2 + install_flag_desc_indent))
    fmt_println "$is_short"
    # trigger ssid
    fmt_println "$is_bold_flags" "[{-${TRIGGER_SSID_FLAG[0]}|--${TRIGGER_SSID_FLAG[1]}} <${TRIGGER_SSID_FLAG[3]}>]" | indent $(($2 + install_flag_desc_indent))
    fmt_println "$is_short" "Specify the trigger ssid. i.e. When this network is around, try to connect to your personal hotspot." | indent $(($2 + install_flag_desc_indent))
    fmt_println "$is_short" "default: connect to your hotspot whenever network changes" | indent $(($2 + install_flag_desc_indent))
    fmt_println "$is_short"
    # retry duration
    fmt_println "$is_bold_flags" "[{-${RETRY_DURATION_FLAG[0]}|--${RETRY_DURATION_FLAG[1]}} <${RETRY_DURATION_FLAG[3]}>]]" | indent $(($2 + install_flag_desc_indent))
    fmt_println "$is_short" "Maximum time duration in seconds to try when trigger conditions are met and not yet connected to hotspot." | indent $(($2 + install_flag_desc_indent))
    fmt_println "$is_short" "default: 600s" | indent $(($2 + install_flag_desc_indent))
    fmt_println "$is_short"
    # uninstall
    fmt_println "$is_bold_flags" "[{-${UNINSTALL_FLAG[0]}|--${UNINSTALL_FLAG[1]}}]" | indent "$2"
    fmt_println "$is_short" "Uninstall this script" | indent "$2"
    fmt_println "$is_short"
    # help
    fmt_println "$is_bold_flags" "[{-${HELP_FLAG[0]}|--${HELP_FLAG[1]}}]" | indent "$2"
    fmt_println "$is_short" "Show help" | indent "$2"
    fmt_println "$is_short"
    # version
    fmt_println "$is_bold_flags" "[{-${VERSION_FLAG[0]}|--${VERSION_FLAG[1]}}]" | indent "$2"
    fmt_println "$is_short" "The version of this script" | indent "$2"
    fmt_println "$is_short"

    # examples
    fmt_println "$is_short" "$BOLD" "============== EXAMPLES ==========================" | indent "$2"
    fmt_println "$is_short" "auto-hotspot.sh -t TRIGGER_SSID -s iPhone" | indent "$2"
    fmt_println
}

# This is a bit of hacking way in that the default value in this plist
# is reserved. i.e. it can occur only once in the correct field.
# Limitation of shell
replace_str_in_plist() {
    sed -i -e "s/$1/$2/g" ./com.nick.auto-hotspot.plist.new
}

check_privilege() {
    if [[ $EUID -ne 0 ]]; then
        echo "This script must be granted elevated privilege to install/uninstall, i.e. sudo $1"
        exit 1
    fi
}

install() {
    cp ./com.nick.auto-hotspot.plist ./com.nick.auto-hotspot.plist.new
    replace_str_in_plist "HOTSPOT_SSID" "$1"

    if [ -n "$2" ]; then
        replace_str_in_plist "TRIGGER_SSID" "$2"
    else
        replace_str_in_plist "TRIGGER_SSID" ""
    fi

    if [ -n "$3" ]; then
        replace_str_in_plist "RETRY_DURATION" "$3"
    else
        replace_str_in_plist "RETRY_DURATION" 600
    fi

    user_name=$(who am i | awk '{print $1}')
    user_id=$(id -u "$user_name")
    replace_str_in_plist "USER_ID" "$user_id"

    cp ./auto-hotspot.sh /Library/Scripts/
    chmod 755 /Library/Scripts/auto-hotspot.sh
    mv ./com.nick.auto-hotspot.plist.new /Library/LaunchDaemons/com.nick.auto-hotspot.plist
    chmod 600 /Library/LaunchDaemons/com.nick.auto-hotspot.plist
    launchctl bootstrap system /Library/LaunchDaemons/com.nick.auto-hotspot.plist
}

uninstall() {
    launchctl bootout system /Library/LaunchDaemons/com.nick.auto-hotspot.plist
    rm /Library/Scripts/auto-hotspot.sh
    rm /Library/LaunchDaemons/com.nick.auto-hotspot.plist
}

# TODO: next version auto grant security keychain access
# TODO: fix getopts unreachable flags when flags like --eee
# TODO: add alerts for uninstall
# TODO: add check for existence for installation
main() {
    trigger_ssid=""
    hotspot_ssid=""
    retry_duration=""
    to_install=false
    has_args=false
    program_indent=$((${#0} + 1))
    optstr=":-:${HELP_FLAG[2]}${VERSION_FLAG[2]}${HOTSPOT_SSID_FLAG[2]}${TRIGGER_SSID_FLAG[2]}${RETRY_DURATION_FLAG[2]}${INSTALL_FLAG[2]}${UNINSTALL_FLAG[2]}"
    while getopts "$optstr" opt; do
        has_args=true
        # support long options: https://stackoverflow.com/a/28466267/519360
        if [ "$opt" = "-" ]; then     # long opt, reformulate OPT and OPTARG
            opt="${OPTARG%%=*}"       # extract long opt name
            OPTARG="${OPTARG#"$opt"}" # extract long opt arg (may be empty)
            OPTARG="${OPTARG#=}"      # if long opt, remove assigning `=`
        fi
        # display Help
        if is_in "$opt" "${HELP_FLAG[@]}"; then
            help "$0" "$program_indent" "OPTION"
            exit
        # version
        elif is_in "$opt" "${VERSION_FLAG[@]}"; then
            echo "version"
            exit
        # trigger ssid
        elif is_in "$opt" "${TRIGGER_SSID_FLAG[@]}"; then
            trigger_ssid="$OPTARG"
        # hotspot ssid
        elif is_in "$opt" "${HOTSPOT_SSID_FLAG[@]}"; then
            hotspot_ssid="$OPTARG"
        # retry duration
        elif is_in "$opt" "${RETRY_DURATION_FLAG[@]}"; then
            retry_duration="$OPTARG"
        # install
        elif is_in "$opt" "${INSTALL_FLAG[@]}"; then
            check_privilege "$0"
            to_install=true
        # uninstall
        elif is_in "$opt" "${UNINSTALL_FLAG[@]}"; then
            check_privilege "$0"
            uninstall
            exit
        elif [ -n "$OPTARG" ]; then
            echo "Error, Invalid argument $OPTARG"
            exit 1
        else
            echo "This should be unreachable. Debug required." >&2
        fi
    done
    if ! $has_args; then
        help "$0" "$program_indent" "USAGE"
        exit
    fi
    if $to_install; then
        if [ -n "$hotspot_ssid" ]; then
            install "$hotspot_ssid" "$trigger_ssid" "$retry_duration"
        else
            fmt_println "$RED" "To install, hotspot ssid must be provided"
            fmt_println
            help "$0" "$program_indent" "USAGE"
            exit 1
        fi
    fi
}

main "$@"
